src-input/duk_bi_encoding.c

```diff
@@ -231,81 +306,77 @@ DUK_INTERNAL duk_ret_t duk_bi_textdecoder_constructor(duk_context *ctx) {
+	 * initialized explicitly.
+	 */
 	dec_ctx = (duk__decode_context *) duk_push_fixed_buffer(ctx, sizeof(duk__decode_context));
 	dec_ctx->fatal = fatal;
 	dec_ctx->ignore_bom = ignore_bom;
-	duk__utf8_decode_init(dec_ctx);
+	duk__utf8_decode_init(dec_ctx);  /* Initializes remaining fields. */
 
-	duk_push_this(ctx);
-	duk_dup(ctx, -2);
 	duk_put_prop_string(ctx, -2, "\xff" "Context");
 	return 0;
 }
 
-DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_encoding_getter(duk_context *ctx) {
+/* Get TextDecoder context from 'this'; leaves garbage on stack. */
+DUK_LOCAL duk__decode_context *duk__get_textdecoder_context(duk_context *ctx) {
 	duk__decode_context *dec_ctx;
-
-	duk_push_this(ctx);
-	duk_get_prop_string(ctx, -1, "\xff" "Context");
-	dec_ctx = (duk__decode_context *) duk_require_buffer(ctx, -1, NULL);
-	DUK_ASSERT(dec_ctx != NULL);
-
-	/* XXX: change to look up from dec_ctx once other encodings are supported */
-	DUK_UNREF(dec_ctx);
-	duk_push_string(ctx, "utf-8");
-	return 1;
-}
-
-DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_fatal_getter(duk_context *ctx) {
-	duk__decode_context *dec_ctx;
-
 	duk_push_this(ctx);
 	duk_get_prop_string(ctx, -1, "\xff" "Context");
 	dec_ctx = (duk__decode_context *) duk_require_buffer(ctx, -1, NULL);
 	DUK_ASSERT(dec_ctx != NULL);
-
-	duk_push_boolean(ctx, dec_ctx->fatal);
-	return 1;
+	return dec_ctx;
 }
 
-DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_ignorebom_getter(duk_context *ctx) {
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_shared_getter(duk_context *ctx) {
 	duk__decode_context *dec_ctx;
+	duk_int_t magic;
+
+	dec_ctx = duk__get_textdecoder_context(ctx);
+	magic = duk_get_current_magic(ctx);
+	switch (magic) {
+	case 0:
+		/* Encoding is now fixed, so _Context lookup is only needed to
+		 * validate the 'this' binding (TypeError if not TextDecoder-like).
+		 */
+		duk_push_string(ctx, "utf-8");
+		break;
+	case 1:
+		duk_push_boolean(ctx, dec_ctx->fatal);
+		break;
+	default:
+		duk_push_boolean(ctx, dec_ctx->ignore_bom);
+		break;
+	}
 
-	duk_push_this(ctx);
-	duk_get_prop_string(ctx, -1, "\xff" "Context");
-	dec_ctx = (duk__decode_context *) duk_require_buffer(ctx, -1, NULL);
-
-	duk_push_boolean(ctx, dec_ctx->ignore_bom);
 	return 1;
 }
 
 DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_decode(duk_context *ctx) {
-	const duk_uint8_t *input = (const duk_uint8_t *) "";
+	const duk_uint8_t *input;
 	duk__decode_context *dec_ctx;
 	duk_size_t len = 0;
 	duk_bool_t stream = 0;
 	duk_codepoint_t codepoint;
-	int ret;
 	duk_uint8_t *output;
 	const duk_uint8_t *in;
 	duk_uint8_t *out;
 
+	dec_ctx = duk__get_textdecoder_context(ctx);
+
 	if (!duk_is_undefined(ctx, 0)) {
 		input = (const duk_uint8_t *) duk_require_buffer_data(ctx, 0, &len);
+	} else {
+		input = NULL;  /* NULL is OK if 'len' is 0. */
+		DUK_ASSERT(len == 0);
 	}
 	if (!duk_is_null_or_undefined(ctx, 1)) {
 		if (duk_get_prop_string(ctx, 1, "stream")) {
```

Comment:
By the way, there's this internal helper for this just case:

    DUK_INTERNAL_DECL duk_hobject *duk_require_hobject_accept_mask(duk_context *ctx, duk_idx_t idx, duk_uint_t type_mask);

It will accept an object (returning its pointer) *or* accept any types in the accept mask (returning NULL).